server/sonar-server/src/main/java/org/sonar/server/computation/filemove/FileMoveDetectionStep.java

```diff
@@ -0,0 +1,462 @@
+        if (fileSourceDto != null) {
+          builder.put(fileKey, new File(componentDto.path(), fileSourceDto.getSrcHash(), on('\n').splitToList(fileSourceDto.getLineHashes())));
+        }
+      }
+      return builder.build();
+    }
+  }
+
+  private Map<String, File> getReportFileSourcesByKey(Map<String, Component> reportFilesByKey, Set<String> addedFileKeys) {
+    ImmutableMap.Builder<String, File> builder = ImmutableMap.builder();
+    for (String fileKey : addedFileKeys) {
+      // FIXME computation of sourceHash and lineHashes might be done multiple times for some files: here, in ComputeFileSourceData, in
+      // SourceHashRepository
+      Component component = reportFilesByKey.get(fileKey);
+      SourceLinesHashesComputer linesHashesComputer = new SourceLinesHashesComputer();
+      SourceHashComputer sourceHashComputer = new SourceHashComputer();
+      try (CloseableIterator<String> lineIterator = sourceLinesRepository.readLines(component)) {
+        while (lineIterator.hasNext()) {
+          String line = lineIterator.next();
+          linesHashesComputer.addLine(line);
+          sourceHashComputer.addLine(line, lineIterator.hasNext());
+        }
+      }
+      builder.put(fileKey, new File(component.getReportAttributes().getPath(), sourceHashComputer.getHash(), linesHashesComputer.getLineHashes()));
+    }
+    return builder.build();
+  }
+
+  private ScoreMatrix computeScoreMatrix(Map<String, File> dbFileSourcesByKey, Map<String, File> reportFileSourcesByKey) {
+    int[][] scoreMatrix = new int[dbFileSourcesByKey.size()][reportFileSourcesByKey.size()];
+    int maxScore = 0;
+
+    int dbFileIndex = 0;
+    for (Map.Entry<String, File> dbFileSourceAndKey : dbFileSourcesByKey.entrySet()) {
+      File fileInDb = dbFileSourceAndKey.getValue();
+      int reportFileIndex = 0;
+      for (Map.Entry<String, File> reportFileSourceAndKey : reportFileSourcesByKey.entrySet()) {
+        File unmatchedFile = reportFileSourceAndKey.getValue();
+        int score = fileSimilarity.score(fileInDb, unmatchedFile);
+        scoreMatrix[dbFileIndex][reportFileIndex] = score;
+        if (score > maxScore) {
+          maxScore = score;
+        }
+        reportFileIndex++;
+      }
+      dbFileIndex++;
+    }
+
+    return new ScoreMatrix(dbFileSourcesByKey, reportFileSourcesByKey, scoreMatrix, maxScore);
+  }
+
+  private static void printIfDebug(ScoreMatrix scoreMatrix) {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("ScoreMatrix:\n" + scoreMatrix.toCsv(';'));
+    }
+  }
+
+  private static ElectedMatches electMatches(Map<String, File> dbFileSourcesByKey, Map<String, File> reportFileSourcesByKey, MatchesByScore matchesByScore) {
+    ElectedMatches electedMatches = new ElectedMatches(matchesByScore, dbFileSourcesByKey, reportFileSourcesByKey);
+    Multimap<String, Match> matchesPerFileForScore = ArrayListMultimap.create();
+    for (List<Match> matches : matchesByScore) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS135)